### PR TITLE
Add CLIP text model

### PR DESCRIPTION
**Pull request by sogartar**
Ports the CLIP text model from Hugging Face. This is the first iteration so not much is changed from the original model. Things like dropout and checkpointing are removed.
Add numeric verification tests for the various components of the stack when executing in eager mode. Verifications are made for float32 and bfloat16. There are tests for toy-sized components and the whole model as well as the Large pretrained variant.
These tests does not include testing with IREE.

Functionalities for mask creation are not yet ported.

### CODE REVIEW

File: sharktank/sharktank/layers/norm.py [hunk 1]
```diff
@@ -39,3 +39,24 @@ def forward(self, x: torch.Tensor):
...
+        weight_name: str = "weight",
+        bias_name: str = "bias",
+        eps: float = 1e-05,
+    ):
+        super().__init__(theta)
+        self.weight = self.theta_tensor(weight_name)
+        if bias_name in self.theta.keys:
+            self.bias = self.theta_tensor(bias_name)
+        else:
+            self.bias = None
```
Comment (rsuderman) [inline sharktank/sharktank/layers/norm.py:58]: Don't bother if-else ing. Just set `self.bias = None` prior to the if statement

File: sharktank/sharktank/models/clip/clip.py [hunk 2]
```diff
@@ -0,0 +1,502 @@
...
+            theta=theta("token_embedding"), dtype=None
+        )
+        self.position_embedding = TokenEmbeddingLayer(
+            theta=theta("position_embedding"), dtype=None
+        )
+
+        # position_ids (1, len position emb) is contiguous in memory and exported when serialized
+        self.register_buffer(
+            "position_ids",
+            torch.arange(config.max_position_embeddings).expand((1, -1)),
```
Comment (rsuderman) [inline sharktank/sharktank/models/clip/clip.py:41]: You should use `unsqueeze` instead of `expand`

File: sharktank/sharktank/models/clip/clip.py [hunk 3]
```diff
@@ -0,0 +1,502 @@
...
+        super().__init__()
+        self.embed_dim = config.hidden_size
+        self.num_heads = config.num_attention_heads
+        self.head_dim = self.embed_dim // self.num_heads
+        if self.head_dim * self.num_heads != self.embed_dim:
+            raise ValueError(
+                f"embed_dim must be divisible by num_heads (got `embed_dim`: {self.embed_dim} and `num_heads`:"
+                f" {self.num_heads})."
+            )
+        self.scale = self.head_dim**-0.5
```
Comment (rsuderman) [inline sharktank/sharktank/models/clip/clip.py:80]: Avoid using `**` for inverse powers. Its better to do `1.0 / math.sqrt(self.head_dim)`. The numerical precision on `pow` operators is usually significantly worse.

File: sharktank/sharktank/models/clip/clip.py [hunk 4]
```diff
@@ -0,0 +1,502 @@
...
+        attn_weights = ops.matmul(query_states, key_states.transpose(1, 2))
+
+        if attn_weights.size() != (bsz * self.num_heads, tgt_len, src_len):
+            raise ValueError(
+                f"Attention weights should be of size {(bsz * self.num_heads, tgt_len, src_len)}, but is"
+                f" {attn_weights.size()}"
+            )
+
+        # apply the causal_attention_mask first
+        if causal_attention_mask is not None:
```
Comment (rsuderman) [inline sharktank/sharktank/models/clip/clip.py:125]: Why do you separate the causal attention mask from the regular attention mask? They should just occur together. Even the causal attention mask should really just be a bool.

File: sharktank/sharktank/models/clip/clip.py [hunk 5]
```diff
@@ -0,0 +1,502 @@
...
+                raise ValueError(
+                    f"Attention mask should be of size {(bsz, 1, tgt_len, src_len)}, but is {attention_mask.size()}"
+                )
+            attn_weights = (
+                attn_weights.view(bsz, self.num_heads, tgt_len, src_len)
+                + attention_mask
+            )
+            attn_weights = attn_weights.view(bsz * self.num_heads, tgt_len, src_len)
+
+        attn_weights = ops.softmax(attn_weights, dim=-1)
```
Comment (rsuderman) [inline sharktank/sharktank/models/clip/clip.py:148]: Rather than decomposing we should use the `ops.scaled_dot_product_attention` operation. Attention is attention, so we should avoid replicating the decomposed version everywhere.

File: sharktank/sharktank/models/clip/clip.py [hunk 6]
```diff
@@ -0,0 +1,502 @@
...
+        output_hidden_states = (
+            output_hidden_states
+            if output_hidden_states is not None
+            else self.config.output_hidden_states
+        )
+        return_dict = (
+            return_dict if return_dict is not None else self.config.use_return_dict
+        )
+
+        encoder_states = () if output_hidden_states else None
```
Comment (rsuderman) [inline sharktank/sharktank/models/clip/clip.py:316]: This logic of empty tuple vs `none` and adding feels unclear what it actually is attempting to do. Relying on `none + tuple` feels weird

File: sharktank/sharktank/types/tensors.py [hunk 7]
```diff
@@ -378,6 +378,12 @@ def reshape(self, *args: Union[List[List[int]], List[int]]) -> "AnyTensor":
             shape = args[0]
         return reshape(self, shape)
 
+    def size(self, dim: Optional[int] = None) -> tuple[int]:
+        if dim is None:
+            return tuple(self.shape)
+        else:
```
Comment (rsuderman) [inline sharktank/sharktank/types/tensors.py:384]: No `else` condition required. Just include the return when the condition is not taken.

**Review by rsuderman — CHANGES_REQUESTED** [review]

File: sharktank/sharktank/models/clip/clip.py (hunk 2 above)
Comment (sogartar) [reply sharktank/sharktank/models/clip/clip.py:41]: I forgot to mention in the PR description that this is the initial port of the model and I did not attempt to optimize anything. The main goal was to put it under test. I would rather do these modifications later as they would be tracked in different commits and will be clear what changed compared to the original.

File: sharktank/sharktank/layers/norm.py (hunk 1 above)
Comment (sogartar) [reply sharktank/sharktank/layers/norm.py]: Done.

File: sharktank/sharktank/types/tensors.py (hunk 7 above)
Comment (sogartar) [reply sharktank/sharktank/types/tensors.py]: Done.

File: sharktank/sharktank/models/clip/clip.py [hunk 8]
```diff
@@ -0,0 +1,502 @@
...
+        if inputs_embeds is None:
+            inputs_embeds = self.token_embedding(input_ids)
+
+        position_embeddings = self.position_embedding(position_ids)
+        embeddings = inputs_embeds + position_embeddings
+
+        return embeddings
+
+
+class ClipAttention(BaseLayer):
```
Comment (IanNod) [inline sharktank/sharktank/models/clip/clip.py:67]: At a quick glance is seems much of this is reused from the punet Attention (outside the decomposed SDPA which we should try avoiding) from here https://github.com/nod-ai/shark-ai/blob/5f98e7321564eeec6d0901250699938c7307ef65/sharktank/sharktank/models/punet/layers.py#L390. Can we just extend and reuse that implementation?

File: sharktank/sharktank/models/clip/clip.py [hunk 9]
```diff
@@ -0,0 +1,502 @@
...
+            return_dict=return_dict,
+        )
+
+        last_hidden_state = (
+            encoder_outputs["last_hidden_state"] if return_dict else encoder_outputs[0]
+        )
+        last_hidden_state = self.final_layer_norm(last_hidden_state)
+
+        if self.eos_token_id == 2:
+            # The `eos_token_id` was incorrect before PR #24773: Let's keep what have been done here.
```
Comment (IanNod) [inline sharktank/sharktank/models/clip/clip.py:427]: This PR I assume is reference to diffusers? Confusing to have in our repo

File: sharktank/tests/models/clip/clip_test.py [hunk 10]
```diff
@@ -0,0 +1,449 @@
...
+@pytest.mark.usefixtures("path_prefix")
+class ClipExportTest(TempDirTestBase):
+    def setUp(self):
+        super().setUp()
+        if self.path_prefix is None:
+            self.path_prefix = f"{self._temp_dir}/"
+
+    @with_clip_data
+    def testSmokeExportLargeF32FromHuggingFace(self):
+        repo_id = "openai/clip-vit-large-patch14"
```
Comment (IanNod) [inline sharktank/tests/models/clip/clip_test.py:60]: We might want to avoid downloading full models for developers running locally. Should we implement a toy model to accommodate? Thoughts @rsuderman?

File: sharktank/sharktank/models/clip/clip.py (hunk 2 above)
Comment (KyleHerndon) [reply sharktank/sharktank/models/clip/clip.py:41]: If you're directly referencing the huggingface code for implementation, do you want to link it in a comment?

File: sharktank/tests/models/clip/clip_test.py (hunk 10 above)
Comment (sogartar) [reply sharktank/tests/models/clip/clip_test.py:60]: Because it downloads data I marked it with `with_clip_data`.

File: sharktank/sharktank/models/clip/clip.py (hunk 2 above)
Comment (sogartar) [reply sharktank/sharktank/models/clip/clip.py:45]: I added a comment at the top of the file.

File: sharktank/sharktank/models/clip/clip.py (hunk 9 above)
Comment (sogartar) [reply sharktank/sharktank/models/clip/clip.py]: I removed this EOS variant.

File: sharktank/sharktank/models/clip/clip.py (hunk 8 above)
Comment (sogartar) [reply sharktank/sharktank/models/clip/clip.py:71]: I would like to iterate on improving code reuse and performance in subsequent PRs.

File: sharktank/sharktank/models/clip/clip.py (hunk 9 above)
Comment (sogartar) [reply sharktank/sharktank/models/clip/clip.py]: We actually go through this case in the actual model so I put it back in, but changed the PR reference to a full link.

**Review by KyleHerndon — APPROVED** [review]

File: sharktank/sharktank/models/clip/clip.py (hunk 3 above)
Comment (sogartar) [reply sharktank/sharktank/models/clip/clip.py:84]: I will address this in a later PR.

File: sharktank/sharktank/models/clip/clip.py (hunk 4 above)
Comment (sogartar) [reply sharktank/sharktank/models/clip/clip.py:129]: This is coming from the original model in Hugging Face. I will address this in a later PR.

File: sharktank/sharktank/models/clip/clip.py (hunk 5 above)
Comment (sogartar) [reply sharktank/sharktank/models/clip/clip.py:152]: This is coming from the original model in Hugging Face. I will address this in a later PR as the whole Layer should get merged with the one from Punet.

File: sharktank/sharktank/models/clip/clip.py (hunk 6 above)
Comment (sogartar) [reply sharktank/sharktank/models/clip/clip.py:320]: I would probably substitute this mechanism with tensor tracing in the future.
